### PR TITLE
Link and Unlink Subnet to an interface

### PR DESCRIPTION
**Pull request by howbazaar**

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -109,6 +125,138 @@ func (i *interface_) Delete() error {
...
+			case http.StatusServiceUnavailable:
+				return errors.Wrap(err, NewCannotCompleteError(svrErr.BodyMessage))
+			}
+		}
+		return NewUnexpectedError(err)
+	}
+
+	response, err := readInterface(i.controller.apiVersion, source)
+	if err != nil {
+		return errors.Trace(err)
```
Comment (mjs) [inline interface.go:215]: Should this be wrapped by one of the error types?

File: interface.go [hunk 2]
```diff
@@ -109,6 +125,138 @@ func (i *interface_) Delete() error {
...
+			case http.StatusForbidden:
+				return errors.Wrap(err, NewPermissionError(svrErr.BodyMessage))
+			}
+		}
+		return NewUnexpectedError(err)
+	}
+
+	response, err := readInterface(i.controller.apiVersion, source)
+	if err != nil {
+		return errors.Trace(err)
```
Comment (mjs) [inline interface.go:256]: As above

File: interface.go [hunk 3]
```diff
@@ -109,6 +125,138 @@ func (i *interface_) Delete() error {
...
+	}
+	var link Link
+	for _, k := range i.links {
+		if s := k.Subnet(); s != nil && s.ID() == subnet.ID() {
+			link = k
+			break
+		}
+	}
+	if link == nil {
+		return errors.NotValidf("unlinked Subnet")
```
Comment (mjs) [inline interface.go:237]: is this the right error text? An error here means that the subnet is linked to this interface but the error message implies the subnet isn't linked at all.

File: interface.go [hunk 4]
```diff
@@ -109,6 +125,138 @@ func (i *interface_) Delete() error {
...
+	i.updateFrom(response)
+	return nil
+}
+
+// LinkSubnet implements Interface.
+func (i *interface_) UnlinkSubnet(subnet Subnet) error {
+	if subnet == nil {
+		return errors.NotValidf("missing Subnet")
+	}
+	var link Link
```
Comment (mjs) [inline interface.go]: I think extracting the subnet search to it's own method would enhance clarity.

**Comment by mjs** [comment]
LGTM - just little things

File: interface.go (hunk 1 above)
Comment (howbazaar) [inline interface.go:215]: No, because the errors returned out of readInterface are already one of the local types.

File: interface.go (hunk 3 above)
Comment (howbazaar) [inline interface.go:237]: If we couldn't find a link with the specified subnet, then the error text is "unlinked Subnet not valid".

It isn't valid to unlink a subnet that isn't linked. Alternatively, we could just succeed ?

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**